C++ API handle optimizer defaults

stmcgovern opened the pull request:
Fixes #141884 

This fixes the issue for all optimizers and parameter options. 
A member function `overwrite_from` is added to the optimizer base class. Each optimizer then implements this function for comparing their accepted parameters to defaults. A SFINAE approach to handle the different optimizer parameters generically (in optimizer.h only) was evaluated, but I think this is easier to review and maintain.

This mirrors the Python API up to one edge case. An example of the edge case is provided below. 

Python can distinguish between 1) Key not present in dict = "not specified"  and 2) Key present in dict = "explicitly set". The C++ implementation cannot.
The issue hinges on whether or not to track if a particular parameter was set by the user explicitly or not (discrepancy in the case when the constructor default is explicitly passed in). 

To track this seems like it will take more intervention than would be worth it (modify TORCH_ARG to keep track, use std::optional for the parameter types, use bitset tracking) and was not pursued in the current PR. I'm happy to alter the design if appropriate. 

### Example of edge case hinging on CONSTRUCTOR DEFAULTS vs OPTIMIZER DEFAULTS 

1. CONSTRUCTOR DEFAULTS:
   These are the values you get when calling AdamOptions()
   AdamOptions().lr() = 0.001
   AdamOptions().weight_decay() = 0
   AdamOptions().eps() = 1e-08

2. OPTIMIZER DEFAULTS:
   These are the values the user chose when creating the optimizer
   User's optimizer defaults:
   optimizer.lr() = 0.005
   optimizer.weight_decay() = 0.1
   optimizer.eps() = 1e-07

3. THE PROBLEM SCENARIO:
   User wants to add a parameter group with explicit weight_decay=0.0
   User sets: weight_decay(0)

4. THE CONFUSION:
   Constructor default weight_decay: 0
   User's explicit weight_decay:     0
   Are they equal? YES

   Since they're equal, our overwrite_from() logic thinks:
   "User didn't set weight_decay explicitly, use optimizer default"

5. CURRENT BEHAVIOR:
   Final weight_decay: 0.1
   User expected:      0
   Match? ❌ NO

=== KEY INSIGHT ===
Constructor defaults are built into the C++ class definition.
Optimizer defaults are chosen by the user at runtime. We want to respect the user intention.

stmcgovern commented:
Hi @janeyx99, this is the PR I mentioned in issue #141884 . How can I link the PR to the issue? I thought the "Fixes #number" does that...

janeyx99 reviewed (commented):
Thanks for working on the fix, this is an interesting bug indeed. I left a comment on the overall approach. Furthermore, since these tests don't run on CI--could you post a paste of the C++ test results locally?

stmcgovern commented on test/cpp/api/optim.cpp:
```diff
@@ -564,3 +566,332 @@ TEST(OptimTest, CheckLRChange_ReduceLROnPlateau_Adam) {
+}
+
+TEST(OptimTest, MergeWithDefaultOptions_AdamW) {
+  torch::manual_seed(0);
```
right. removed

stmcgovern commented on torch/csrc/api/src/optim/optimizer.cpp:
```diff
@@ -94,17 +94,32 @@ void OptimizerOptions::serialize(
       "You must override it in your subclass of torch::optim::OptimizerCloneableOptions<YourOptimizerOptions>.");
 }
 
+void OptimizerOptions::overwrite_from(const OptimizerOptions& source) {
```
we don't :) . I've tried to provide a fix to the base class, without adding a new API.

stmcgovern commented:
Hi @janeyx99, thanks very much for your feedback. I've taken each one of your comments into consideration.
I've tried to find a complete fix to the base class, without adding a new API. The core problem is that we need a way to track whether the parameter field was explicitly set or not (to mimic Python dict behavior). I started with a function pointer registry and string based checking in add_param_group. Trying to move as much to compile time as possible, 1) led to the CRTP pattern for static dispatch and 2) avoiding string hash checking, with the bitset tracking, was adopted for performance and ease of serialization. 
This uses a new macro (which I also don't like), but seems to solve the problem in an efficient and complete way (python api parity), without introducing a new API.

Here are the local optimizer tests (including the new ones). Is there a better way with gtest to see what the test actually ran (not just pass/fail)? dropping the filter and running 1020 tests pass too.

```
(base) [root@49023e5b8e19 pytorch]# ./build/bin/test_api --gtest_filter="*Optim*" -v
CUDA not available. Disabling CUDA and MultiCUDA tests
Note: Google Test filter = *Optim*-*_CUDA:*_MultiCUDA
[==========] Running 51 tests from 2 test suites.
[----------] Global test environment set-up.
[----------] 44 tests from OptimTest
[ RUN      ] OptimTest.OptimizerAccessors
[       OK ] OptimTest.OptimizerAccessors (1 ms)
[ RUN      ] OptimTest.OldInterface
[       OK ] OptimTest.OldInterface (0 ms)
[ RUN      ] OptimTest.XORConvergence_SGD
[       OK ] OptimTest.XORConvergence_SGD (655 ms)
[ RUN      ] OptimTest.XORConvergence_LBFGS
[       OK ] OptimTest.XORConvergence_LBFGS (434 ms)
[ RUN      ] OptimTest.XORConvergence_Adagrad
[       OK ] OptimTest.XORConvergence_Adagrad (250 ms)
[ RUN      ] OptimTest.XORConvergence_RMSprop
[       OK ] OptimTest.XORConvergence_RMSprop (236 ms)
[ RUN      ] OptimTest.XORConvergence_RMSpropWithMomentum
[       OK ] OptimTest.XORConvergence_RMSpropWithMomentum (703 ms)
[ RUN      ] OptimTest.XORConvergence_Adam
[       OK ] OptimTest.XORConvergence_Adam (264 ms)
[ RUN      ] OptimTest.XORConvergence_AdamWithAmsgrad
[       OK ] OptimTest.XORConvergence_AdamWithAmsgrad (278 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_Adam
[       OK ] OptimTest.ProducesPyTorchValues_Adam (92 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_AdamWithWeightDecay
[       OK ] OptimTest.ProducesPyTorchValues_AdamWithWeightDecay (96 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_AdamWithWeightDecayAndAMSGrad
[       OK ] OptimTest.ProducesPyTorchValues_AdamWithWeightDecayAndAMSGrad (99 ms)
[ RUN      ] OptimTest.XORConvergence_AdamW
[       OK ] OptimTest.XORConvergence_AdamW (267 ms)
[ RUN      ] OptimTest.XORConvergence_AdamWWithAmsgrad
[       OK ] OptimTest.XORConvergence_AdamWWithAmsgrad (266 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_AdamW
[       OK ] OptimTest.ProducesPyTorchValues_AdamW (94 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_AdamWWithoutWeightDecay
[       OK ] OptimTest.ProducesPyTorchValues_AdamWWithoutWeightDecay (95 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_AdamWWithAMSGrad
[       OK ] OptimTest.ProducesPyTorchValues_AdamWWithAMSGrad (99 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_Adagrad
[       OK ] OptimTest.ProducesPyTorchValues_Adagrad (78 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_AdagradWithWeightDecay
[       OK ] OptimTest.ProducesPyTorchValues_AdagradWithWeightDecay (83 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_AdagradWithWeightDecayAndLRDecay
[       OK ] OptimTest.ProducesPyTorchValues_AdagradWithWeightDecayAndLRDecay (83 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_RMSprop
[       OK ] OptimTest.ProducesPyTorchValues_RMSprop (82 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_RMSpropWithWeightDecay
[       OK ] OptimTest.ProducesPyTorchValues_RMSpropWithWeightDecay (86 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_RMSpropWithWeightDecayAndCentered
[       OK ] OptimTest.ProducesPyTorchValues_RMSpropWithWeightDecayAndCentered (94 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_RMSpropWithWeightDecayAndCenteredAndMomentum
[       OK ] OptimTest.ProducesPyTorchValues_RMSpropWithWeightDecayAndCenteredAndMomentum (99 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_SGD
[       OK ] OptimTest.ProducesPyTorchValues_SGD (69 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_SGDWithWeightDecay
[       OK ] OptimTest.ProducesPyTorchValues_SGDWithWeightDecay (71 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_SGDWithWeightDecayAndMomentum
[       OK ] OptimTest.ProducesPyTorchValues_SGDWithWeightDecayAndMomentum (72 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_SGDWithWeightDecayAndNesterovMomentum
[       OK ] OptimTest.ProducesPyTorchValues_SGDWithWeightDecayAndNesterovMomentum (80 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_LBFGS
[       OK ] OptimTest.ProducesPyTorchValues_LBFGS (68 ms)
[ RUN      ] OptimTest.ProducesPyTorchValues_LBFGS_with_line_search
[       OK ] OptimTest.ProducesPyTorchValues_LBFGS_with_line_search (298 ms)
[ RUN      ] OptimTest.ZeroGrad
[       OK ] OptimTest.ZeroGrad (0 ms)
[ RUN      ] OptimTest.ExternalVectorOfParameters
[       OK ] OptimTest.ExternalVectorOfParameters (0 ms)
[ RUN      ] OptimTest.AddParameter_LBFGS
[       OK ] OptimTest.AddParameter_LBFGS (0 ms)
[ RUN      ] OptimTest.CheckLRChange_StepLR_Adam
[       OK ] OptimTest.CheckLRChange_StepLR_Adam (0 ms)
[ RUN      ] OptimTest.CheckLRChange_ReduceLROnPlateau_Adam
[       OK ] OptimTest.CheckLRChange_ReduceLROnPlateau_Adam (0 ms)
[ RUN      ] OptimTest.MergeWithDefaultOptions_Adam
[       OK ] OptimTest.MergeWithDefaultOptions_Adam (0 ms)
[ RUN      ] OptimTest.MergeWithDefaultOptions_SGD
[       OK ] OptimTest.MergeWithDefaultOptions_SGD (0 ms)
[ RUN      ] OptimTest.MergeWithDefaultOptions_AdamW
[       OK ] OptimTest.MergeWithDefaultOptions_AdamW (0 ms)
[ RUN      ] OptimTest.MergeWithDefaultOptions_Adagrad
[       OK ] OptimTest.MergeWithDefaultOptions_Adagrad (0 ms)
[ RUN      ] OptimTest.MergeWithDefaultOptions_RMSprop
[       OK ] OptimTest.MergeWithDefaultOptions_RMSprop (0 ms)
[ RUN      ] OptimTest.MergeWithDefaultOptions_LBFGS
[       OK ] OptimTest.MergeWithDefaultOptions_LBFGS (0 ms)
[ RUN      ] OptimTest.MergeWithDefaultOptions_NoOptionsInheritance
[       OK ] OptimTest.MergeWithDefaultOptions_NoOptionsInheritance (0 ms)
[ RUN      ] OptimTest.SerializationPreservesFieldTracking_Adam
[       OK ] OptimTest.SerializationPreservesFieldTracking_Adam (9 ms)
[ RUN      ] OptimTest.SerializationPreservesFieldTracking_SGD
[       OK ] OptimTest.SerializationPreservesFieldTracking_SGD (0 ms)
[----------] 44 tests from OptimTest (5217 ms total)

[----------] 7 tests from SerializeTest
[ RUN      ] SerializeTest.Optim
[       OK ] SerializeTest.Optim (1 ms)
[ RUN      ] SerializeTest.Optim_Adagrad
[       OK ] SerializeTest.Optim_Adagrad (1 ms)
[ RUN      ] SerializeTest.Optim_SGD
[       OK ] SerializeTest.Optim_SGD (1 ms)
[ RUN      ] SerializeTest.Optim_Adam
[       OK ] SerializeTest.Optim_Adam (1 ms)
[ RUN      ] SerializeTest.Optim_AdamW
[       OK ] SerializeTest.Optim_AdamW (1 ms)
[ RUN      ] SerializeTest.Optim_RMSprop
[       OK ] SerializeTest.Optim_RMSprop (1 ms)
[ RUN      ] SerializeTest.Optim_LBFGS
[       OK ] SerializeTest.Optim_LBFGS (1 ms)
[----------] 7 tests from SerializeTest (12 ms total)

[----------] Global test environment tear-down
[==========] 51 tests from 2 test suites ran. (5230 ms total)
[  PASSED  ] 51 tests.
(base) [root@49023e5b8e19 pytorch]# 
```

stmcgovern commented:
Thanks for your comments @janeyx99 . That makes sense! I certainly agree that users should not have to be aware of these implementation details. I wasn't sure how much the runtime performance cost impacted your review. will revisit and simplify.

stmcgovern committed:
[C++ API] Implement parameter group inheritance for optimizer parity

Addresses PyTorch issue #141884 by implementing automatic parameter group
inheritance that achieves Python-C++ API parity without breaking changes.

- Uses comparison-based merging to infer user intent vs default inheritance
- C++17 SFINAE patterns following PyTorch conventions (matches c10/util/TypeTraits.h)
-Add comprehensive tests for optimizer parameter group inheritance

stmcgovern commented:
Hi @janeyx99, here is my preferred approach. It does everything in optimizer.h/cpp, doesn't introduce any new API and most work is done at compile time. I've tried to follow how c10 uses template metaprogramming style. C++20 concepts can help smooth out some of the  boilerplate, when they become available in PyTorch.  Local tests are passing. Please let me know what you think. Thanks!

janeyx99 reviewed (commented):
Much nicer! Can we privatize all the helpers?

stmcgovern committed:
prefix private helper functions with underscore

stmcgovern commented on torch/csrc/api/include/torch/optim/optimizer.h:
```diff
@@ -51,19 +52,47 @@ class OptimizerCloneableParamState : public OptimizerParamState {
+
+ protected:
+  void mark_field_as_set(size_t field_id) {
+    if (field_id < 32) {
```
outdated

stmcgovern commented on torch/csrc/api/include/torch/optim/optimizer.h:
```diff
@@ -66,12 +67,314 @@ class TORCH_API OptimizerOptions {
+      }
+    }
+    if constexpr (has_tolerance_grad<Derived>::value) {
+      if (user_options.tolerance_grad() !=
```
great questions. 
1) yes it's always set - either as user specified or as a constructor default (which won't get used, since the derived optimizer defaults will be selected in the merge process).  Unspecified fields in user_options contain constructor defaults, but we don't use them aside from comparison.

2) if we always set the result->attr to user_options.attr(), then we would end up with constructor defaults in the final parameters. This is not what the user expects. User intent would be that when overriding a default, it does not affect the values of the unspecified fields (which should be the optimizer defaults, not the constructor defaults). That matches the python api expectations.

janeyx99 commented on torch/csrc/api/include/torch/optim/optimizer.h:
```diff
@@ -66,12 +67,314 @@ class TORCH_API OptimizerOptions {
 };
 
-template <typename Derived>
+// Forward declarations for optimizer option types
```
Can we make the following classes and structs private as well?

stmcgovern replied:
since these are forward declarations, I'm inclined to not change the style to use the prefix (would need touching all optimizer files). I do have to change them to be struct (which resolves the inconsistency causing clang build failure).

janeyx99 commented on torch/csrc/api/include/torch/optim/optimizer.h:
```diff
@@ -66,12 +67,314 @@ class TORCH_API OptimizerOptions {
+
+  // SFINAE field detection - detects optimizer fields using public accessor methods
+  template <class T, class Enable = void>
+  struct has_lr : std::false_type {};
```
These structs too

janeyx99 reviewed (dismissed):
Please do privatize as much as possible so we are not inadvertently growing our API surface.

stmcgovern committed:
use _ pattern for sfinae structs and fix fwd declaration consistency

stmcgovern committed:
fix lint

stmcgovern commented:
OK I think this is ready. Thanks for your feedback and help @janeyx99 !

amjames commented:
@pytorchbot merge -r main

amjames commented:
@pytorchbot merge

stmcgovern committed:
fix linker vtable errors
